R/learn_cvpairs.R

```diff
@@ -1,37 +1,71 @@
 #' Cross-validation pairs
 #'
-#' Add cross validation pairs to a pipelearner object. \code{n} random
-#' partitions will try to be created unless \code{k} is not NULL, in which case
-#' k folds will be created. See \code{\link[modelr]{crossv_mc}} for more
-#' details.
+#' Add cross validation pairs to a pipelearner object. By default will create
+#' a random partition of the data. Can use any function which returns a data
+#' data frame with three columns: \code{train}, \code{test}, and \code{.id}.
+#' \code{test} and \code{train} are list-columns with each entry of type
+#' \code{resample}. \code{.id} is an atomic vector. If no function is specified,
+#' random partitions will try to be created unless \code{k} is not NULL, in
+#' which case k folds will be created. See \code{\link[modelr]{crossv_mc}} for
+#' more details.
 #'
 #' Will expect a pipelearner object, but will also accept a data frame for the
 #' parameter \code{pl}. In the case that a data frame is used, it will first
 #' coerce it to a pipelearner object via \code{\link{pipelearner}}.
 #'
 #' @inheritParams pipelearner_params
-#' @inheritParams modelr::crossv_mc
+#' @param .f a function which takes a \code{data} argument
+#' @param ... additional parameters to be passed to \code{.f}
+#'
+#' @examples
+#' pl <- pipelearner(mtcars)
+#' learn_cvpairs(pl, crossv_kfold, k = 4)
 #' @export
-learn_cvpairs <- function(pl, k = NULL, n = 1, test = 0.2) {
+learn_cvpairs <- function(pl, .f, ...) {
   UseMethod("learn_cvpairs")
 }
 
 #' @export
-learn_cvpairs.default <-  function(pl, k = NULL, n = 1, test = 0.2) {
-  learn_cvpairs.pipelearner(pipelearner(pl), k = k, n = n, test = test)
+learn_cvpairs.default <-  function(pl, .f, ...) {
+  learn_cvpairs.pipelearner(pipelearner(pl), ...)
 }
 
 #' @export
-learn_cvpairs.pipelearner <- function(pl, k = NULL, n = 1, test = 0.2) {
-
-  # n random partitions
-  if (is.null(k)) {
-    cv_pairs <- modelr::crossv_mc(pl$data, n = n, test = test, id = ".id")
+learn_cvpairs.pipelearner <- function(pl, .f, ...) {
```

Comment:
This function has become a bit unwieldy. I am happy to fix it to my liking after merging, but there appear to be some more serious issues. Specifically, the default case throws an error. E.g., try running `pipelearner(mtcars, lm, mpg ~ .)` and you'll get an error straight away.

Either way, for default case, I had something simpler in mind. Like:

```
if (missing(.f)) {
  pl$cv_pairs <- crossv_mc(pl$data, n = 1, test = 0.2, id = ".id")
  return (pl)
}
```

See if this helps with errors.